Implement CICD identical to algorithms-jupyter-extension

frankinspace opened the pull request:
Overview of work done:
- Implemented CICD which re-uses the workflows defined in https://github.com/MAAP-Project/algorithms-jupyter-extension/tree/develop/.github/workflows
- Package name needed to be changed to use `_` instead of `-` which is consistent with https://github.com/MAAP-Project/algorithms-jupyter-extension/blob/develop/package.json#L2
- Refactored tests to match the file structure of https://github.com/jupyterlab/extension-examples/tree/main/hello-world/ui-tests and https://github.com/MAAP-Project/algorithms-jupyter-extension/tree/develop/ui-tests

Published release from `feature/cicd` branch:
- Github Action run: https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/runs/14226893488
- Pypi published: https://pypi.org/project/maap-che-sidebar-visibility-jupyter-extension/1.1.3/
- GitHub release: https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/releases/tag/v1.1.3

Commit:
New build workflow

Commit:
New build workflow

Commit:
bump version of @jupyterlab/builder

Commit:
build errors

Commit:
build errors

Commit:
build errors

Commit:
build errors

Commit:
build errors

Commit:
build errors

Commit:
Trying to get the tests working

Commit:
name needs _ not -

Commit:
name needs _ not -

Commit:
Cleanup for pr

Commit:
Cleanup for pr

Commit:
Cleanup for pr

frankinspace committed:
Merge branch 'develop' into feature/cicd

grallewellyn commented:
Has this been tested to simulate a release? I think we should test since you will likely be off the project for the next release 
Maybe we can just increase the version by a 0.0.1 since it will publish to PyPi

Commit:
unquote boolean

Commit:
parse boolean input

Commit:
remove unsupported options

frankinspace commented:
@grallewellyn I'm in progress on a test release of 1.1.3; just waiting on a request to authorize the CICD app for this repository (and the other *-extension repos).

grallewellyn commented:
@frankinspace Do I need to approve this request? I didn't see one

frankinspace commented:
> @frankinspace Do I need to approve this request? I didn't see one

@grallewellyn not yet, I was still waiting for the CICD app auth updates. Now that has been granted I ran into another issue.

Turns out pypi does not support using a reusable workflow in order to setup trusted publishing (https://github.com/pypi/warehouse/issues/11096)
So that means the CICD workflow and the Step 1: Prepare Release workflows can be shared but Step 2: Publish Release workflow will need to be defined in each *-jupyter-extension repo until that issue is closed and released to pypi.

So I'll make that change and try publishing again and the re-request review.

Commit:
Calling a reusable workflow to publish to pypi will not work until https://github.com/pypi/warehouse/issues/11096 is resolved

frankinspace committed:
Publish 1.1.3

SHA256 hashes:

maap_che_sidebar_visibility_jupyter_extension-1.1.3-py3-none-any.whl: 80707c31597d8572fa13e7b5f1e6b43692419f3b7bfd27788e4b0179d6a09055

maap_che_sidebar_visibility_jupyter_extension-1.1.3.tar.gz: dcd766c63fd2f735e3ea5309339842b02486522ea73d9f4562f953ecb2691440

maap_che_sidebar_visibility_jupyter_extension-1.1.3.tgz: 1e6d6ff506b48063f565c28a952d2436441377bbabe0b1daeaa4e94ffa920ab2

frankinspace commented:
v1.1.3 was successfully published from the `feature/cicd` branch

https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/runs/14226893488
https://pypi.org/project/maap-che-sidebar-visibility-jupyter-extension/1.1.3/
https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/releases/tag/v1.1.3

@MAAP-Project/platform this is ready for review

grallewellyn commented:
@frankinspace Just to confirm, the process when we do a release for this would be:

```
Develop branch is on v4.2.0a0
Merge feature 1 into develop:
- This creates a pre release for the tag 4.2.0a1
- This publishes 4.2.0a1 to test pypi

Time for release 
Branch release/4.2.1 (you choose this name) off of the develop branch  (This will create v4.2.1-rc0 tag)
- Realize bug in this release, merge fix into this release branch (This will create v4.2.1-rc1)
- This will be published to test PyPi by the CI/CD with version v4.2.1-rc1?? 
- Pre release tag 4.2.1-rc1 on GitHub 

Merging PR release branch into main
This creates 4.2.1 on PyPi 
4.2.1 as a tag and the latest release on GitHub 
Then need to do PR back from main to develop. Easiest to do this from the commandline and accept the develop version 
```

> This will be published to test PyPi by the CI/CD with version v4.2.1-rc1?? 

This step is the only one I would like input on

grallewellyn commented:
@frankinspace Can you please verify the process I described above? I can approve your PRs then

frankinspace commented:
@grallewellyn The jupyter extension repos use a slightly different CICD pipeline because of the jupyter-releaser tool.

So these repositories will not auto-version themselves on commits. Instead you need to run the Step 1 / Step 2 actions in order to do the release. One of the inputs to Step 1 is `New Version Specifier` which you can set to a specific version number or leave it as next to allow hatch to handle the version bump

grallewellyn commented:
@frankinspace What step 1 and step 2 are you referring to?

frankinspace commented:
@grallewellyn These two actions:
https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/workflows/prep-release.yml
https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/workflows/publish-release.yml

grallewellyn commented:
Could you please revise this process I sent: 
```
Develop branch is on v4.2.0a0
Merge feature 1 into develop:
- This creates a pre release for the tag 4.2.0a1
- This publishes 4.2.0a1 to test pypi

Time for release 
Branch release/4.2.1 (you choose this name) off of the develop branch  (This will create v4.2.1-rc0 tag)
- Realize bug in this release, merge fix into this release branch (This will create v4.2.1-rc1)
- This will be published to test PyPi by the CI/CD with version v4.2.1-rc1?? 
- Pre release tag 4.2.1-rc1 on GitHub 

Merging PR release branch into main
This creates 4.2.1 on PyPi 
4.2.1 as a tag and the latest release on GitHub 
Then need to do PR back from main to develop. Easiest to do this from the commandline and accept the develop version 
```

With where and when I am supposed to be running the workflow for https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/workflows/prep-release.yml and 
https://github.com/MAAP-Project/che-sidebar-visibility-jupyter-extension/actions/workflows/publish-release.yml?

What should we do for `The branch to target`? Or `Use PRs with activity since this date or git reference` because there is also the box to click this? 
It is easier if this process is written out instead of in a video for the person doing the release

frankinspace commented:
Since we decided to try using the Jupyter Releaser for these repos the process is different. Really, the release process for the extensions does not need to change at all from what was happening previously. The new functionality is that:

1. Build and tests will run when commits are pushed to tracked branches (currently `develop, release/**, main, master, feature/**, issue/**, issues/**, dependabot/**`). The version number will NOT be changed by the CI process
2. When a release is ready to be published, run Step 1: Prep Release. This will create a draft release in github and update the changelog to include activity that has occurred since the last release. The "Branch to target" is the name of the branch that the release is being created from. "Use PRs with activity since this date or git reference" lets you exclude some PRs/commits if needed but I expect will remain blank in nearly all cases. The "Use PRs with activity since the last stable git tag" checkbox is similar but will include changes based on the last tag. Again, I'd recommend just leaving unchecked and letting it default to including activity since the last release.
3. After the draft release is created it can be reviewed and tweaks can be made. If all looks good then Step 2: Publish Release can be used to publish the github release, create the tag, and publish the artifacts to PyPi.

More documentation is available here: https://jupyter-releaser.readthedocs.io/en/latest/

I can update the MAAP Release Process google doc to capture these two new steps.

grallewellyn commented:
Thanks! Should 

> The "Branch to target" is the name of the branch that the release is being created from

Be like `release/x.y.z`?
